discoverd/main.go

```diff
@@ -260,6 +278,106 @@ func waitHostDNSConfig() (addr string, resolvers []string) {
 	return addr, status.Network.Resolvers
 }
 
+func inPeerList(peers []string, addr string) bool {
+	for _, peer := range peers {
+		if peer == addr {
+			return true
+		}
+	}
+	return false
+}
+
+// Join the consensus set, promoting ourselves from proxy to raft node.
+func (m *Main) Promote() error {
+	m.logger.Printf("attempting promotion")
+	if m.store != nil {
+		return fmt.Errorf("can't join, already a raft peer")
+	}
+	if !m.handler.Proxy.Load().(bool) {
+		return fmt.Errorf("can't promote, handler not in proxy mode")
+	}
+
+	// Obtain leader address from one of the peers we know about.
+	var leaderAddr string
+	for _, peer := range m.peers {
+		client := discoverd.NewClientWithURL(peer)
+		leader, err := client.RaftLeader()
+		if err != nil {
+			continue
+		}
+		leaderAddr = leader.Host
+		break
+	}
+	if leaderAddr == "" {
+		return fmt.Errorf("failed to get leader address from configured peers")
+	}
+
+	// Request the leader joins us to the cluster.
+	m.logger.Println("requesting leader join us to cluster")
+	if err := discoverd.NewClientWithURL(leaderAddr).RaftAddPeer(m.advertiseAddr); err != nil {
+		m.logger.Println("error requesting leader to join us to cluster:", err)
+		return err
+	}
+
+	// Open the store.
+	if err := m.openStore(); err != nil {
+		return err
+	}
+
+	// Wait for leadership.
+	if err := m.waitForLeader(LeaderTimeout); err != nil {
+		return err
+	}
+
+	// Update the DNS server to use the local store.
+	m.dnsServer.SetStore(m.store)
+
+	// Update the HTTP server to use the local store.
+	m.handler.Store = m.store
+	m.handler.Proxy.Store(false)
+
+	m.logger.Println("promoted sucessfully")
+	return nil
+}
+
+// Leave the consensus set, demoting ourselves to proxy from raft node.
+func (m *Main) Demote() error {
+	m.logger.Println("demotion requested")
+	if m.store == nil {
+		return fmt.Errorf("can't leave, not currently a raft peer")
+	}
+	if m.handler.Proxy.Load().(bool) {
+		return fmt.Errorf("can't demote, handler in proxy mode")
+	}
+
+	// Update the dns server to use proxy store
+	m.dnsServer.SetStore(&server.ProxyStore{Peers: m.peers})
+	m.handler.Proxy.Store(true)
+
+	// If we are the leader remove ourselves directly,
+	// otherwise request the leader remove us from the peer set.
+	leaderAddr := m.store.Leader()
+	if leaderAddr == "" {
+		return server.ErrNoKnownLeader
```

Comment:
In this case and other error cases below the handler and DNS server will not get reset back to the original, is that okay?